### PR TITLE
[JUJU-3786] Added controller config domain

### PR DESCRIPTION
**Pull request by anvial**
This PR introduces the controller config domain, which includes the necessary state API and services to facilitate operations with controller configurations stored in the database. 

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
~- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
~- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```sh
cd domain/controllerconfig/service
go test -check.f serviceSuite

cd ../state
go test -check.f stateSuite

cd ..
go test -check.f controllerconfigSuite
```

### CODE REVIEW

File: domain/controllerconfig/state/state.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+func (st *State) ControllerConfig(ctx context.Context) (jujucontroller.Config, error) {
+	return nil, nil
+}
+
+// UpdateControllerConfig allows changing some of the configuration
+// for the controller. Changes passed in updateAttrs will be applied
+// to the current config, and keys in removeAttrs will be unset (and
+// so revert to their defaults). Only a subset of keys can be changed
+// after bootstrapping.
+func (st *State) UpdateControllerConfig(ctx context.Context, updateAttrs map[string]interface{}, removeAttrs []string) error {
```
Comment (manadart) [inline domain/controllerconfig/state/state.go:36]: Interested to see how we can use [this feature](https://github.com/canonical/sqlair/pull/48) when it lands.

File: domain/controllerconfig/state/state.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		}
+
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+// checkUpdateControllerConfig checks that the given name is a valid.
+func checkUpdateControllerConfig(name string) error {
```
[the remark this answers is not in the document]
Comment (anvial) [reply domain/controllerconfig/state/state.go]: done

File: domain/controllerconfig/state/state.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+// ControllerConfig returns the current configuration in the database.
+func (st *State) ControllerConfig(ctx context.Context) (jujucontroller.Config, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	q := `
```
Comment (manadart) [inline domain/controllerconfig/state/state.go:35]: Just stick it on one line and forget the indexing.

File: domain/controllerconfig/state/state.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Validate the updateAttrs.
+	fields, _, err := jujucontroller.ConfigSchema.ValidationSchema()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for k := range updateAttrs {
```
Comment (manadart) [inline domain/controllerconfig/state/state.go:67]: Do the coercion in the service layer and pass a map to this method.

File: domain/controllerconfig/state/state.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+		}
+
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+// controllerConfigFromRows returns controller config info from rows returned from the backing DB.
+func controllerConfigFromRows(rows *sql.Rows) (jujucontroller.Config, error) {
```
Comment (manadart) [inline domain/controllerconfig/state/state.go:106]: Return a map and keep coercion in the service layer.

**Review by manadart — COMMENTED** [review]
Some suggested changes.

File: domain/controllerconfig/service/service.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// UpdateControllerConfig updates the controller config.
+func (s *Service) UpdateControllerConfig(ctx context.Context, updateAttrs jujucontroller.Config, removeAttrs []string) error {
+	err := s.st.UpdateControllerConfig(ctx, updateAttrs, removeAttrs)
+	return errors.Annotate(err, "updating controller config state")
+}
+
+// checkUpdateControllerConfig checks that the given name is a valid.
+func checkUpdateControllerConfig(name string) error {
```
Comment (manadart) [inline domain/controllerconfig/service/service.go:47]: You are not calling this.

**Review by manadart — COMMENTED** [review]
Still a few things to tidy up.

**Review by manadart — APPROVED** [review]
Nice one.

**Pull request merged**